testlapack/dgerqf.go

```diff
@@ -0,0 +1,95 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/floats"
+)
+
+type Dgerqfer interface {
+	Dgerq2er
+	Dgerqf(m, n int, a []float64, lda int, tau, work []float64, lwork int)
+}
+
+func DgerqfTest(t *testing.T, impl Dgerqfer) {
+	rnd := rand.New(rand.NewSource(1))
+	for c, test := range []struct {
+		m, n, lda int
+	}{
+		{10, 5, 0},
+		{5, 10, 0},
+		{10, 10, 0},
+		{300, 5, 0},
+		{3, 500, 0},
+		{200, 200, 0},
+		{300, 200, 0},
+		{204, 300, 0},
+		{1, 3000, 0},
+		{3000, 1, 0},
+		{10, 5, 20},
+		{5, 10, 20},
+		{10, 10, 20},
+		{300, 5, 400},
+		{3, 500, 600},
+		{200, 200, 300},
+		{300, 200, 300},
+		{204, 300, 400},
+		{1, 3000, 4000},
+		{3000, 1, 4000},
+	} {
+		m := test.m
+		n := test.n
+		lda := test.lda
+		if lda == 0 {
+			lda = test.n
+		}
+		a := make([]float64, m*lda)
+		for i := 0; i < m; i++ {
+			for j := 0; j < n; j++ {
+				a[i*lda+j] = rnd.Float64()
+			}
+		}
+		tau := make([]float64, n)
+		for i := 0; i < n; i++ {
+			tau[i] = rnd.Float64()
+		}
+		aCopy := make([]float64, len(a))
+		copy(aCopy, a)
+		ans := make([]float64, len(a))
+		copy(ans, a)
+		work := make([]float64, m)
+		// Compute unblocked QR.
+		impl.Dgerq2(m, n, ans, lda, tau, work)
```

Comment:
This is elegant and reasonable but as always the cgo version cannot be tested well with this.